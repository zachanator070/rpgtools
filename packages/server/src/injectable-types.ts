export const INJECTABLE_TYPES = {
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PermissionAssignmentRepository: Symbol.for("PermissionAssignmentRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	FileRepository: Symbol.for("FileRepository"),

	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	SessionContextFactory: Symbol.for("SessionContextFactory"),
	Open5eApiClient: Symbol.for("Open5eApiClient"),

	Cache: Symbol.for("Cache"),
};
