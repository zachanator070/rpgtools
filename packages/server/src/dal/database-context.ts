import {RepositoryAccessor} from "../types";
import {inject, injectable} from "inversify";
import {ArticleRepository} from "./repository/article-repository";
import {GameRepository} from "./repository/game-repository";
import {FileRepository} from "./repository/file-repository";
import {ChunkRepository} from "./repository/chunk-repository";
import {ImageRepository} from "./repository/image-repository";
import {ItemRepository} from "./repository/item-repository";
import {ModelRepository} from "./repository/model-repository";
import {MonsterRepository} from "./repository/monster-repository";
import {PersonRepository} from "./repository/person-repository";
import {PinRepository} from "./repository/pin-repository";
import {PlaceRepository} from "./repository/place-repository";
import {RoleRepository} from "./repository/role-repository";
import {ServerConfigRepository} from "./repository/server-config-repository";
import {UserRepository} from "./repository/user-repository";
import {WikiFolderRepository} from "./repository/wiki-folder-repository";
import {WikiPageRepository} from "./repository/wiki-page-repository";
import {WorldRepository} from "./repository/world-repository";
import {INJECTABLE_TYPES} from "../di/injectable-types";
import {DatabaseSession} from "./database-session";

@injectable()
export class DatabaseContext implements RepositoryAccessor {

    @inject(INJECTABLE_TYPES.ArticleRepository)
    articleRepository: ArticleRepository;
    @inject(INJECTABLE_TYPES.ChunkRepository)
    chunkRepository: ChunkRepository;
    @inject(INJECTABLE_TYPES.FileRepository)
    fileRepository: FileRepository;
    @inject(INJECTABLE_TYPES.GameRepository)
    gameRepository: GameRepository;
    @inject(INJECTABLE_TYPES.ImageRepository)
    imageRepository: ImageRepository;
    @inject(INJECTABLE_TYPES.ItemRepository)
    itemRepository: ItemRepository;
    @inject(INJECTABLE_TYPES.ModelRepository)
    modelRepository: ModelRepository;
    @inject(INJECTABLE_TYPES.MonsterRepository)
    monsterRepository: MonsterRepository;
    @inject(INJECTABLE_TYPES.PersonRepository)
    personRepository: PersonRepository;
    @inject(INJECTABLE_TYPES.PinRepository)
    pinRepository: PinRepository;
    @inject(INJECTABLE_TYPES.PlaceRepository)
    placeRepository: PlaceRepository;
    @inject(INJECTABLE_TYPES.RoleRepository)
    roleRepository: RoleRepository;
    @inject(INJECTABLE_TYPES.ServerConfigRepository)
    serverConfigRepository: ServerConfigRepository;
    @inject(INJECTABLE_TYPES.UserRepository)
    userRepository: UserRepository;
    @inject(INJECTABLE_TYPES.WikiFolderRepository)
    wikiFolderRepository: WikiFolderRepository;
    @inject(INJECTABLE_TYPES.WikiPageRepository)
    wikiPageRepository: WikiPageRepository;
    @inject(INJECTABLE_TYPES.WorldRepository)
    worldRepository: WorldRepository;

    databaseSession: DatabaseSession;

}