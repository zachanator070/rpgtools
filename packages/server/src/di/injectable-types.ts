import {DatabaseContext} from "../dal/database-context";

export const INJECTABLE_TYPES = {

	DomainEntity: Symbol.for("DomainEntity"),

	// domain entities
	Article: Symbol.for("Article"),
	Chunk: Symbol.for("Chunk"),
	File: Symbol.for("File"),
	Game: Symbol.for("Game"),
	Image: Symbol.for("Image"),
	Item: Symbol.for("Item"),
	Model: Symbol.for("Model"),
	Monster: Symbol.for("Monster"),
	PermissionAssignment: Symbol.for("PermissionAssignment"),
	Person: Symbol.for("Person"),
	Pin: Symbol.for("Pin"),
	Place: Symbol.for("Place"),
	Role: Symbol.for("Role"),
	ServerConfig: Symbol.for("ServerConfig"),
	User: Symbol.for("User"),
	WikiFolder: Symbol.for("WikiFolder"),
	World: Symbol.for("World"),
	Event: Symbol.for("Event"),
	Calendar: Symbol.for("Calendar"),

	// entity factories
	ArticleFactory: Symbol.for("ArticleFactory"),
	AclFactory: Symbol.for('AclFactory'),
	ChunkFactory: Symbol.for("ChunkFactory"),
	FileFactory: Symbol.for("FileFactory"),
	GameFactory: Symbol.for("GameFactory"),
	ImageFactory: Symbol.for("ImageFactory"),
	ItemFactory: Symbol.for("ItemFactory"),
	ModelFactory: Symbol.for("ModelFactory"),
	MonsterFactory: Symbol.for("MonsterFactory"),
	PersonFactory: Symbol.for("PersonFactory"),
	PinFactory: Symbol.for("PinFactory"),
	PlaceFactory: Symbol.for("PlaceFactory"),
	RoleFactory: Symbol.for("RoleFactory"),
	ServerConfigFactory: Symbol.for("ServerConfigFactory"),
	UserFactory: Symbol.for("UserFactory"),
	WikiFolderFactory: Symbol.for("WikiFolderFactory"),
	WikiPageFactory: Symbol.for("WikiPageFactory"),
	WorldFactory: Symbol.for("WorldFactory"),
	EventWikiFactory: Symbol.for("EventFactory"),
	CalendarFactory: Symbol.for("CalendarFactory"),

	CharacterFactory: Symbol.for('CharacterFactory'),
	CharacterAttributeFactory: Symbol.for('CharacterAttributeFactory'),
	PathNodeFactory: Symbol.for('NodePathFactory'),
	StrokeFactory: Symbol.for('StrokeFactory'),
	FogStrokeFactory: Symbol.for('FogStrokeFactory'),
	MessageFactory: Symbol.for('MessageFactory'),
	InGameModelFactory: Symbol.for('InGameModelFactory'),

	AgeFactory: Symbol.for("AgeFactory"),
	MonthFactory: Symbol.for("MonthFactory"),
	DayOfTheWeekFactory: Symbol.for("DayOfTheWeekFactory"),

	// repositories
	ArticleRepository: Symbol.for("ArticleRepository"),
	ChunkRepository: Symbol.for("ChunkRepository"),
	FileRepository: Symbol.for("FileRepository"),
	GameRepository: Symbol.for("GameRepository"),
	ImageRepository: Symbol.for("ImageRepository"),
	ItemRepository: Symbol.for("ItemRepository"),
	ModelRepository: Symbol.for("ModelRepository"),
	MonsterRepository: Symbol.for("MonsterRepository"),
	PersonRepository: Symbol.for("PersonRepository"),
	PinRepository: Symbol.for("PinRepository"),
	PlaceRepository: Symbol.for("PlaceRepository"),
	RoleRepository: Symbol.for("RoleRepository"),
	ServerConfigRepository: Symbol.for("ServerConfigRepository"),
	UserRepository: Symbol.for("UserRepository"),
	WikiFolderRepository: Symbol.for("WikiFolderRepository"),
	WikiPageRepository: Symbol.for("WikiPageRepository"),
	WorldRepository: Symbol.for("WorldRepository"),
	EventWikiRepository: Symbol.for("EventWikiRepository"),
	CalendarRepository: Symbol.for("CalendarRepository"),
	StrokeRepository: Symbol.for('StrokeRepository'),
	FogStrokeRepository: Symbol.for('FogStrokeRepository'),

	SqlPermissionControlledRepository: Symbol.for('SqlPermissionControlledRepository'),

	// authorization policies
	ArticleAuthorizationPolicy: Symbol.for("ArticleAuthorizationPolicy"),
	ChunkAuthorizationPolicy: Symbol.for("ChunkAuthorizationPolicy"),
	FileAuthorizationPolicy: Symbol.for("FileAuthorizationPolicy"),
	GameAuthorizationPolicy: Symbol.for("GameAuthorizationPolicy"),
	ImageAuthorizationPolicy: Symbol.for("ImageAuthorizationPolicy"),
	ItemAuthorizationPolicy: Symbol.for("ItemAuthorizationPolicy"),
	ModelAuthorizationPolicy: Symbol.for("ModelAuthorizationPolicy"),
	MonsterAuthorizationPolicy: Symbol.for("MonsterAuthorizationPolicy"),
	PersonAuthorizationPolicy: Symbol.for("PersonAuthorizationPolicy"),
	PinAuthorizationPolicy: Symbol.for("PinAuthorizationPolicy"),
	PlaceAuthorizationPolicy: Symbol.for("PlaceAuthorizationPolicy"),
	RoleAuthorizationPolicy: Symbol.for("RoleAuthorizationPolicy"),
	ServerConfigAuthorizationPolicy: Symbol.for("ServerConfigAuthorizationPolicy"),
	UserAuthorizationPolicy: Symbol.for("UserAuthorizationPolicy"),
	WikiFolderAuthorizationPolicy: Symbol.for("WikiFolderAuthorizationPolicy"),
	WikiPageAuthorizationPolicy: Symbol.for("WikiPageAuthorizationPolicy"),
	WorldAuthorizationPolicy: Symbol.for("WorldAuthorizationPolicy"),
	CalendarAuthorizationPolicy: Symbol.for("CalendarAuthorizationPolicy"),
	FogStrokeAuthorizationPolicy: Symbol.for('FogStrokeAuthorizationPolicy'),
	StrokeAuthorizationPolicy: Symbol.for('StrokeAuthorizationPolicy'),

	// archive repositories
	ArchiveArticleRepository: Symbol.for("ArchiveArticleRepository"),
	ArchiveChunkRepository: Symbol.for("ArchiveChunkRepository"),
	ArchiveGameRepository: Symbol.for("ArchiveGameRepository"),
	ArchiveImageRepository: Symbol.for("ArchiveImageRepository"),
	ArchiveItemRepository: Symbol.for("ArchiveItemRepository"),
	ArchiveModelRepository: Symbol.for("ArchiveModelRepository"),
	ArchiveMonsterRepository: Symbol.for("ArchiveMonsterRepository"),
	ArchivePermissionAssignmentRepository: Symbol.for("ArchivePermissionAssignmentRepository"),
	ArchivePersonRepository: Symbol.for("ArchivePersonRepository"),
	ArchivePinRepository: Symbol.for("ArchivePinRepository"),
	ArchivePlaceRepository: Symbol.for("ArchivePlaceRepository"),
	ArchiveRoleRepository: Symbol.for("ArchiveRoleRepository"),
	ArchiveServerConfigRepository: Symbol.for("ArchiveServerConfigRepository"),
	ArchiveUserRepository: Symbol.for("ArchiveUserRepository"),
	ArchiveWikiFolderRepository: Symbol.for("ArchiveWikiFolderRepository"),
	ArchiveWikiPageRepository: Symbol.for("ArchiveWikiPageRepository"),
	ArchiveWorldRepository: Symbol.for("ArchiveWorldRepository"),
	ArchiveFileRepository: Symbol.for("ArchiveFileRepository"),

	// services
	AuthenticationService: Symbol.for("AuthenticationService"),
	AuthorizationService: Symbol.for("AuthorizationService"),
	ContentExportService: Symbol.for("ContentExportService"),
	ContentImportService: Symbol.for("ContentImportService"),
	GameService: Symbol.for("GameService"),
	ImageService: Symbol.for("ImageService"),
	ModelService: Symbol.for("ModelService"),
	RoleService: Symbol.for("RoleService"),
	ServerConfigService: Symbol.for("ServerConfigService"),
	SrdImportService: Symbol.for("SrdImportService"),
	UserService: Symbol.for("UserService"),
	WikiFolderService: Symbol.for("WikiFolderService"),
	WikiPageService: Symbol.for("WikiPageService"),
	WorldService: Symbol.for("WorldService"),

	// session contexts
	SessionContextFactory: Symbol.for("SessionContextFactory"),
	SecurityContextFactory: Symbol.for("SecurityContextFactory"),

	// 5e
	Open5eApiClient: Symbol.for("Open5eApiClient"),
	Dnd5eApiClient: Symbol.for("Dnd5eApiClient"),

	// delta
	DeltaFactory: Symbol.for("DeltaFactory"),

	// cache
	Cache: Symbol.for("Cache"),

	// archive
	ArchiveFactory: Symbol.for("ArchiveFactory"),

	ZipArchiveFactory: Symbol.for("ZipArchiveFactory"),
	ZipArchive: Symbol.for('ZipArchive'),

	// event publisher
	EventPublisher: Symbol.for("EventPublisher"),

	// server
	ApiServer: Symbol.for("ApiServer"),
	ServerProperties: Symbol.for("ServerProperties"),
	RpgToolsServer: Symbol.for('RpgToolsServer'),
	DbEngine: Symbol.for("DbEngine"),
	SqlDbEngine: Symbol.for('SqlDbEngine'),

	// data loaders
	ArticleDataLoader: Symbol.for("ArticleDataLoader"),
	ChunkDataLoader: Symbol.for("ChunkDataLoader"),
	FileDataLoader: Symbol.for("FileDataLoader"),
	GameDataLoader: Symbol.for("GameDataLoader"),
	ImageDataLoader: Symbol.for("ImageDataLoader"),
	ItemDataLoader: Symbol.for("ItemDataLoader"),
	ModelDataLoader: Symbol.for("ModelDataLoader"),
	MonsterDataLoader: Symbol.for("MonsterDataLoader"),
	PermissionAssignmentDataLoader: Symbol.for("PermissionAssignmentDataLoader"),
	PersonDataLoader: Symbol.for("PersonDataLoader"),
	PinDataLoader: Symbol.for("PinDataLoader"),
	PlaceDataLoader: Symbol.for("PlaceDataLoader"),
	RoleDataLoader: Symbol.for("RoleDataLoader"),
	ServerConfigDataLoader: Symbol.for("ServerConfigDataLoader"),
	UserDataLoader: Symbol.for("UserDataLoader"),
	WikiFolderDataLoader: Symbol.for("WikiFolderDataLoader"),
	WikiPageDataLoader: Symbol.for("WikiPageDataLoader"),
	WorldDataLoader: Symbol.for("WorldDataLoader"),
	CalendarDataLoader: Symbol.for('CalendarDataLoader'),
	EventDataLoader: Symbol.for('EventDataLoader'),

	// seeders
	RoleSeeder: Symbol.for("RoleSeeder"),
	ServerConfigSeeder: Symbol.for("ServerConfigSeeder"),

	// database context
	DatabaseContext: Symbol.for("DatabaseContext"),
	DatabaseContextFactory: Symbol.for("DatabaseContextFactory"),

	// repo mapper
	EntityMapper: Symbol.for("EntityMapper"),

	// mongodb
	FilterFactory: Symbol.for("FilterFactory"),
	MongoDbMigrationV40: Symbol.for("MongoDbMigrationV40"),


};
