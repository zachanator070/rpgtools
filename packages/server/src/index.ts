import {container} from "./di/inversify";
import {INJECTABLE_TYPES} from "./di/injectable-types";
import {ApiServer} from "./types";

const server = container.get<ApiServer>(INJECTABLE_TYPES.ApiServer);

(async () => {
	await server.start();
})();
