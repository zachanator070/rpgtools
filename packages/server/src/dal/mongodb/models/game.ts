import mongoose, {Schema} from "mongoose";
import {GAME, MODEL, PLACE, WIKI_PAGE, WORLD} from "@rpgtools/common/src/type-constants";
import {MongoDBDocument, PermissionControlledDocument} from "../../../types";
import {AclEntry} from "./acl-entry";

const pathNode = new mongoose.Schema({
	x: Number,
	y: Number,
	_id: String,
});

const strokeSchema = new mongoose.Schema({
	path: [pathNode],
	color: {
		type: String,
	},
	size: Number,
	fill: Boolean,
	type: {
		type: String,
		enum: ["circle", "square", "erase", "line"],
	},
	_id: String,
});

const fogStrokeSchema = new mongoose.Schema({
	path: [pathNode],
	size: Number,
	type: {
		type: String,
		enum: ["fog", "erase"],
	},
	_id: String,
});

const gameSchema = new mongoose.Schema({
	passwordHash: {
		type: String,
	},
	world: {
		type: mongoose.Schema.Types.ObjectId,
		ref: WORLD,
		required: [true, "worldId required"],
	},
	map: {
		type: mongoose.Schema.Types.ObjectId,
		ref: PLACE,
	},
	characters: [
		new mongoose.Schema({
			_id: {
				type: String,
				required: [true, "_id required"],
			},
			name: {
				type: String,
			},
			player: {
				type: mongoose.Schema.Types.ObjectId,
				ref: "User",
				required: [true, "player id required"],
			},
			color: {
				type: String,
				required: [true, "color is required"],
			},
			attributes: [
				new mongoose.Schema({
					_id: {
						type: String,
						required: [true, "_id required"],
					},
					name: {
						type: String,
						required: [true, "name required"],
					},
					value: {
						type: Number,
						required: [true, "value required"],
					},
				})
			]
		}),
	],
	host: {
		type: mongoose.Schema.Types.ObjectId,
		ref: "User",
		required: [true, "host id required"],
	},
	strokes: [strokeSchema],
	fog: [fogStrokeSchema],
	messages: [
		new mongoose.Schema({
			_id: {
				type: String,
				required: [true, "_id required"],
			},
			sender: {
				type: String,
				required: [true, "sender required"],
			},
			senderUser: {
				type: String,
				required: [true, "sender user required"],
			},
			receiver: {
				type: String,
				required: [true, "receiver required"],
			},
			receiverUser: {
				type: String,
				required: [true, "receiver user required"],
			},
			message: {
				type: String,
				required: [true, "message required"],
			},
			timestamp: {
				type: String,
				required: [true, "timestamp required"],
			},
		}),
	],
	models: [
		new mongoose.Schema({
			_id: {
				type: String,
				required: [true, "_id required"],
			},
			model: {
				type: mongoose.Schema.Types.ObjectId,
				ref: MODEL,
				required: [true, "model required"],
			},
			x: {
				type: Number,
				required: [true, "x position required"],
			},
			z: {
				type: Number,
				required: [true, "z position required"],
			},
			lookAtX: {
				type: Number,
				required: [true, "x look at point required"],
			},
			lookAtZ: {
				type: Number,
				required: [true, "z look at point required"],
			},
			color: {
				type: String,
			},
			wiki: {
				type: mongoose.Schema.Types.ObjectId,
				ref: WIKI_PAGE,
			},
		}),
	],
	acl: [AclEntry],
});

export interface GameDocument extends MongoDBDocument, PermissionControlledDocument {
	passwordHash: string;
	world: Schema.Types.ObjectId;
	map: Schema.Types.ObjectId;
	characters: CharacterDocument[];
	host: Schema.Types.ObjectId;
	strokes: StrokeDocument[];
	fog: FogStrokeDocument[];
	models: InGameModelDocument[];
	messages: MessageDocument[];
}

export interface PathNodeDocument extends MongoDBDocument {
    x: number;
    y: number;
}

export interface StrokeDocument extends MongoDBDocument {
    path: PathNodeDocument[];
    color: string;
    size: number;
    fill: boolean;
    type: string;
}

export interface FogStrokeDocument extends MongoDBDocument {
    path: PathNodeDocument[];
    size: number;
    type: string;
}

export interface CharacterDocument extends MongoDBDocument {
    name: string;
    player: Schema.Types.ObjectId;
    color: string;
    attributes: CharacterAttributeDocument[];
}

export interface CharacterAttributeDocument extends MongoDBDocument {
    name: string;
    value: number;
}

export interface MessageDocument extends MongoDBDocument {
    sender: string;
    senderUser: string;
    receiver: string;
    receiverUser: string;
    message: string;
    timestamp: number;
}

export interface InGameModelDocument extends MongoDBDocument {
    model: Schema.Types.ObjectId;
    x: number;
    z: number;
    lookAtX: number;
    lookAtZ: number;
    color: string;
    wiki: Schema.Types.ObjectId;
}

export const GameModel = mongoose.model<GameDocument>(GAME, gameSchema);
