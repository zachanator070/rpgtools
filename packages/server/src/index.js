import express from "express";
import bodyParser from "body-parser";
import gqlServer from "./gql-server";
import cookieParser from "cookie-parser";
import morgan from "morgan";
import mongoose from "mongoose";
import { defaultSeeders } from "./default-seeders";
import path from "path";
import { ImageRouter } from "./routers/image-router";
import { checkConfig, serverNeedsSetup } from "./server-needs-setup";
import crypto from "crypto";
import * as http from "http";
import { ServerConfig } from "./models/server-config";
import { ModelRouter } from "./routers/model-router";
import ExportRouter from "./routers/export-router";
import { graphqlUploadExpress } from "graphql-upload";

const mongodb_host = process.env.MONGODB_HOST || "mongodb";
const mongodb_db_name = process.env.MONGODB_DB_NAME || "rpgtools";

mongoose
  .connect(`mongodb://${mongodb_host}/${mongodb_db_name}`, {
    useNewUrlParser: true,
  })
  .then(async () => {
    console.log(
      `Connected to mongodb at mongodb://${mongodb_host}/${mongodb_db_name}`
    );
    await defaultSeeders();
    await createServer();
  });

mongoose.set("useNewUrlParser", true);
mongoose.set("useFindAndModify", false);
mongoose.set("useCreateIndex", true);
mongoose.set("useUnifiedTopology", true);

const createServer = async () => {
  const server = express();

  server.use(bodyParser.json());
  server.use(morgan("tiny"));

  server.use(cookieParser());

  await checkConfig();
  const serverConfig = await ServerConfig.findOne();
  if (serverNeedsSetup()) {
    console.warn(
      `Server needs configuration! Use unlock code ${serverConfig.unlockCode} to unlock`
    );
  }

  server.get("*.js", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/javascript");
    next();
  });
  server.get("*.css", function (req, res, next) {
    req.url = req.url + ".gz";
    res.set("Content-Encoding", "gzip");
    res.set("Content-Type", "text/css");
    next();
  });
  server.get("*favicon.ico", function (req, res, next) {
    req.url = "/favicon.ico";
    next();
  });

  server.use("/images", ImageRouter);
  server.use("/models", ModelRouter);
  server.use("/export", ExportRouter);

  server.get("*", (req, res, next) => {
    const needsSetup = serverNeedsSetup();
    if (needsSetup) {
      if (
        [
          "/api",
          "/ui/setup",
          "/app.bundle.js",
          "/app.css",
          "/app.bundle.js.gz",
          "/app.css.gz",
          "favicon.ico",
        ].includes(req.url)
      ) {
        return next();
      }
      return res.redirect(302, "/ui/setup");
    } else {
      return next();
    }
  });

  server.get("/ui*", (req, res) => {
    return res.sendFile(path.resolve(__dirname, "../../dist", "index.html"));
  });

  server.use(express.static("dist"));
  server.use(express.static("src/static-assets"));
  server.use(graphqlUploadExpress());

  gqlServer.applyMiddleware({ app: server, path: "/api" });

  if (!process.env.ACCESS_TOKEN_SECRET) {
    console.warn(
      "environment variable ACCESS_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.ACCESS_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  if (!process.env.REFRESH_TOKEN_SECRET) {
    console.log(
      "environment variable REFRESH_TOKEN_SECRET is not set, restarting server will log out all users"
    );
    process.env.REFRESH_TOKEN_SECRET = crypto.randomBytes(2048);
  }

  const port = process.env.SERVER_PORT || 3000;

  const httpServer = http.createServer(server);
  gqlServer.installSubscriptionHandlers(httpServer);

  httpServer.listen(port, () => {
    console.log(
      `The server is running and listening at http://localhost:${port}`
    );
  });
};
