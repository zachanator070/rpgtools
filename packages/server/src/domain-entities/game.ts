import {
	AclEntry,
	DomainEntity, EntityFactory,
	PermissionControlledEntity,
	RepositoryAccessor,
} from "../types";
import { GameAuthorizationPolicy } from "../security/policy/game-authorization-policy";
import { GAME } from "@rpgtools/common/src/type-constants";
import { inject, injectable } from "inversify";
import { INJECTABLE_TYPES } from "../di/injectable-types";
import {Repository} from "../dal/repository/repository";
import {GameDocument} from "../dal/mongodb/models/game";
import GameModel from "../dal/sql/models/game-model";

@injectable()
export class Game implements PermissionControlledEntity {
	public _id: string;
	public passwordHash: string;
	public world: string;
	public map?: string;
	public characters: Character[];
	public strokes: Stroke[];
	public fog: FogStroke[];
	public messages: Message[];
	public models: InGameModel[];
	public host: string;
	public acl: AclEntry[];

	authorizationPolicy: GameAuthorizationPolicy;
	factory: EntityFactory<Game, GameDocument, GameModel>;

	type: string = GAME;

	constructor(@inject(INJECTABLE_TYPES.GameAuthorizationPolicy)
					authorizationPolicy: GameAuthorizationPolicy,
				@inject(INJECTABLE_TYPES.GameFactory)
					factory: EntityFactory<Game, GameDocument, GameModel>) {
		authorizationPolicy.entity = this;
		this.authorizationPolicy = authorizationPolicy;
		this.factory = factory;
	}

	getRepository(accessor: RepositoryAccessor): Repository<DomainEntity> {
		return accessor.gameRepository;
	}
}

export class Character {
	public _id: string;
	public name: string;
	public player: string;
	public color: string;
	public attributes: CharacterAttribute[];

	type: string;

	constructor(
		id: string,
		name: string,
		playerId: string,
		color: string,
		attributes: CharacterAttribute[]
	) {
		this._id = id;
		this.name = name;
		this.player = playerId;
		this.color = color;
		this.attributes = attributes;
	}
}

export class CharacterAttribute {
	public _id: string;
	public name: string;
	public value: number;

	constructor(
		_id: string,
		name: string,
		value: number
	) {
		this._id = _id;
		this.name = name;
		this.value = value;
	}
}

export class Stroke {

	public _id: string;
	public path: PathNode[];
	public color: string;
	public size: number;
	public fill: boolean;
	public type: string;

	constructor(
		id: string,
		path: PathNode[],
		color: string,
		size: number,
		fill: boolean,
		type: string
	) {
		this._id = id;
		this.path = path;
		this.color = color;
		this.size = size;
		this.fill = fill;
		this.type = type;
	}
}

export class FogStroke {
	public _id: string;
	public path: PathNode[];
	public size: number;
	public type: string;

	constructor(id: string, path: PathNode[], size: number, type: string) {
		this._id = id;
		this.path = path;
		this.size = size;
		this.type = type;
	}
}

export class PathNode {
	public _id: string;
	public x: number;
	public y: number;

	type: string;

	constructor(id: string, x: number, y: number) {
		this._id = id;
		this.x = x;
		this.y = y;
	}
}

export class Message {
	public sender: string;
	public senderUser: string;
	public receiver: string;
	public receiverUser: string;
	public message: string;
	public timestamp: number;
	public _id: string;

	type: string;

	constructor(sender: string, senderUser: string, receiver: string, receiverUser: string, message: string, timestamp: number, id: string) {
		this.sender = sender;
		this.senderUser = senderUser;
		this.receiver = receiver;
		this.receiverUser = receiverUser;
		this.message = message;
		this.timestamp = timestamp;
		this._id = id;
	}
}

export class InGameModel {
	public _id: string;
	public model: string;
	public x: number;
	public z: number;
	public lookAtX: number;
	public lookAtZ: number;
	public color: string | null;
	public wiki: string | null;

	type: string;

	constructor(
		id: string,
		modelId: string,
		x: number,
		z: number,
		lookAtX: number,
		lookAtZ: number,
		color: string,
		wikiId: string
	) {
		this._id = id;
		this.model = modelId;
		this.x = x;
		this.z = z;
		this.lookAtX = lookAtX;
		this.lookAtZ = lookAtZ;
		this.color = color;
		this.wiki = wikiId;
	}
}
