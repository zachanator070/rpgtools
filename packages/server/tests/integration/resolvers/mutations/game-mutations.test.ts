import {container} from "../../../../src/di/inversify";
import {INJECTABLE_TYPES} from "../../../../src/di/injectable-types";
import {DbEngine, Factory} from "../../../../src/types";
import {DatabaseContext} from "../../../../src/dal/database-context";
import {DefaultTestingContext} from "../../default-testing-context";
import {TEST_INJECTABLE_TYPES} from "../../injectable-types";
import {GameService} from "../../../../src/services/game-service";
import {CREATE_GAME, JOIN_GAME} from "@rpgtools/common/src/gql-mutations";
import {Game} from "../../../../src/domain-entities/game";

process.env.TEST_SUITE = "game-mutations-test";

describe("game mutations", () => {
    const gameService = container.get<GameService>(INJECTABLE_TYPES.GameService);
    const databaseContextFactory = container.get<Factory<DatabaseContext>>(INJECTABLE_TYPES.DatabaseContextFactory);
    const dbEngine = container.get<DbEngine>(INJECTABLE_TYPES.DbEngine);
    const testingContext = container.get<DefaultTestingContext>(TEST_INJECTABLE_TYPES.DefaultTestingContext);


    describe("with world", () => {
        beforeEach(async () => {
            await testingContext.reset();
        });

        describe('as anon', () => {
            beforeEach(async () => {
                testingContext.mockSessionContextFactory.useAnonUser();
            });

            test('create game', async () => {
                const result = await testingContext.server.executeGraphQLQuery({
                    query: CREATE_GAME,
                    variables: { worldId: testingContext.world._id, password: 'testingpassword', username: 'anon' },
                });
                expect(result).toMatchSnapshot({
                    errors: expect.arrayContaining([expect.any(Object)]),
                });
            });
        });

        describe('as tester1 user', () => {
            beforeEach(async () => {
                testingContext.mockSessionContextFactory.setCurrentUser(testingContext.tester1);
            });

            test('create game', async () => {
                const result = await testingContext.server.executeGraphQLQuery({
                    query: CREATE_GAME,
                    variables: { worldId: testingContext.world._id, password: 'testingpassword', username: 'anon' },
                });
                expect(result).toMatchSnapshot({
                    data: {
                        createGame: expect.objectContaining({
                            _id: expect.any(String)
                        })
                    },
                    errors: undefined
                });
            });
        });

        describe('with game hosted by tester1', () => {
            let testGame: Game = null;
            beforeEach(async () => {
                const session = await dbEngine.createDatabaseSession();
                const databaseContext = databaseContextFactory({session});
                testGame = await gameService.createGame(
                    testingContext.tester1SecurityContext,
                    testingContext.world._id,
                    'tester1password',
                    'Tester1',
                    databaseContext
                );
                await session.commit();
            });

            describe('as tester2 user', () => {
                beforeEach(() => {
                    testingContext.mockSessionContextFactory.setCurrentUser(testingContext.tester2);
                });

                test('join existing game', async () => {
                    const result = await testingContext.server.executeGraphQLQuery({
                        query: JOIN_GAME,
                        variables: { gameId: testGame._id, password: 'tester1password', username: 'Tester2' },
                    });
                    expect(result).toMatchSnapshot({
                        data: {
                            joinGame: expect.objectContaining({
                                _id: expect.any(String),
                                characters: [
                                    expect.objectContaining({
                                        _id: expect.any(String),
                                        name: 'Tester1'
                                    }),
                                    expect.objectContaining({
                                        _id: expect.any(String),
                                        name: 'Tester2'
                                    }),
                                ]
                            })
                        },
                        errors: undefined
                    });
                });
            })
        })
    });
});