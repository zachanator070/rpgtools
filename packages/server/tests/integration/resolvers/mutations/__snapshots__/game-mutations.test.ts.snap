// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game mutations with world as anon create game 1`] = `
{
  "data": null,
  "errors": ArrayContaining [
    Any<Object>,
  ],
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;

exports[`game mutations with world as tester1 user create game 1`] = `
{
  "data": {
    "createGame": ObjectContaining {
      "_id": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": {
    "headers": Headers {
      Symbol(map): {},
    },
  },
}
`;
